### PR TITLE
[DOC] Added Joins

### PR DESCRIPTION
**Pull request by svekars**
* Added the Joins concept section.
* Added a Joins example.
* Added description to the pipeline spec.

Co-authored-by: msteffen@pachyderm.io

### CODE REVIEW

File: doc/concepts/pipeline-concepts/pipeline/join.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+if the column names match.
+
+Similarly, joins help you consolidate results with matching names from
+multiple repositories in a single dataset. For example, you collect
+data from multiple sources and need to consolidate the results
+that match a specific parameter. These results can be readings
+from a sensor, a geographical location, or something similar.
+
+When you configure a join input, you must specify a glob pattern
+that defines a capture group. A capture group is the group of
```
[the remark this answers is not in the document]
Comment (JoeyZwicker) [reply doc/concepts/pipeline-concepts/pipeline/join.md]: Also, capture groups isn't something we invented -- pretty sure its a standard unix globbing definition, although certainly a bit archaic. We should look up how they're defined elswhere.

File: doc/concepts/pipeline-concepts/pipeline/join.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+files selected for further processing. In the glob parameter,
```
Comment (JoeyZwicker) [inline doc/concepts/pipeline-concepts/pipeline/join.md]: > In the glob parameter,
a capturing glob must be enclosed in parenthesis. For example,
if you have a file named `test`, you could configure the
following globs for it:

In the glob parameter, a capture group is specified by parentheses. Capture groups are numbered from left to right and can also be nested within each other. You can read more about capture group conventions here (LINK TO UNIX DOCS). Below are a few examples using a file path: `dir123/file456.txt`

`dir123/file(456).txt`  --> capture group 1 = `456`
`(dir123)/file(456).txt`  --> capture group 1 = `123`. capture group 2 = `456`. 
`dir123/file(4(5)6).txt`  --> capture group 1 = `456`. capture group 2 = `5`. 

--- 
started writing this out and realized that this actually needs to get reorganized a bit. I'm happy to take a rewrite pass on it, but I'll get to it tomorrow. Submitting this for now just to not lose track

File: doc/concepts/pipeline-concepts/pipeline/join.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
+# Join
+
+A join is a special type of pipeline input that enables you to combine files that reside in separate Pachyderm repositories and match a particular naming pattern. The join operator must be used in combination with a glob pattern that reflects a specific naming convention.
+
+By analogy, a Pachyderm join is similar to a database equi-join (aka: inner left join) operation, but it's matching on file paths only, not the contents of the files.
```
Comment (adelelopez) [inline doc/concepts/pipeline-concepts/pipeline/join.md]: it's just an inner join

File: examples/joins/README.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+When your repositories are ready, create and run the pipeline
+from the [joins.json]() file by completing the following steps:
+
+1. Run:
+
+   ```bash
+   $ pachctl create pipeline -f joins.json
+   ```
+
```
[the remark this answers is not in the document]
Comment (svekars) [reply examples/joins/README.md]: added

File: examples/joins/README.md [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+1. View the job status by running:
```
[the remark this answers is not in the document]
Comment (svekars) [reply examples/joins/README.md]: added

File: examples/joins/README.md [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+   ```bash
+   $ pachctl list job
+   ID                               PIPELINE STARTED       DURATION           RESTART PROGRESS  DL   UL  STATE
+   7390b0c29ac247c893422a5c04565719 joins    2 seconds ago Less than a second 0       6 + 0 / 6 108B 24B success
+   ```
+
+   In the output above, you can see that Pachyderm processes six datums.
```
[the remark this answers is not in the document]
Comment (svekars) [reply examples/joins/README.md]: there are 6 files.

File: examples/joins/README.md [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+   ```bash
+   $ pachctl list file joins@master
+   NAME       TYPE SIZE
+   /file1.txt file 4B
+   /file2.txt file 4B
+   /file3.txt file 4B
+   /file4.txt file 4B
+   /file5.txt file 4B
+   /file6.txt file 4B
```
[the remark this answers is not in the document]
Comment (svekars) [reply examples/joins/README.md]: 6 files.

File: examples/joins/joins.json [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  "pipeline": {
+    "name": "joins"
+  },
+  "input": {
+    "join": [
+      {
+        "pfs": {
+          "repo": "readings",
+          "branch": "master",
+          "glob": "*/(*).txt",
```
[the remark this answers is not in the document]
Comment (svekars) [reply examples/joins/joins.json]: fixed.

File: examples/joins/joins.json [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+          "join_on": "$1"
+        }
+      },
+     {
+       "pfs": {
+         "repo": "parameters",
+         "branch": "master",
+         "glob": "(*).txt",
```
[the remark this answers is not in the document]
Comment (svekars) [reply examples/joins/joins.json]: fixed.

**Comment by svekars** [comment]
resolved all comments. Ready for review @JoeyZwicker

File: doc/concepts/pipeline-concepts/pipeline/join.md [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+nested capture groups is based on their opening parenthesis.
+
+Below you can find a few examples of applying a glob pattern with a capture
+group to a file path. For example, if you have the following file path:
+
+```bash
+/foo/bar-123/ABC.txt
+```
+
+The following regular expressions, create the following capture groups:
```
Comment (JoeyZwicker) [inline doc/concepts/pipeline-concepts/pipeline/join.md]:
```suggestion
The following glob patterns if used in a join input, would create the following capture groups:
```

File: doc/concepts/pipeline-concepts/pipeline/join.md [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+| Regular expression  | Capture groups           |
+| ------------------- | ------------------------ |
+| `/(*)`              | `foo`                    |
+| `/*/bar-(*)`        | `123`                    |
+| `/(*)/*/(??)*.txt`  | Capture group 1: `foo`, capture group 2: `AB`. |
+| `/*/(bar-(123))/*`  | Capture group 1: `bar-123`, capture group 2: `123`. |
+
+
+Also, joins require you to specify a [replacement group](https://www.regular-expressions.info/replacebackref.html)
+in the  a `join_on`` parameter to define which capture groups you want to try
```
Comment (JoeyZwicker) [inline doc/concepts/pipeline-concepts/pipeline/join.md]:
```suggestion
in the  a `join_on` parameter to define which capture groups you want to try
```

File: examples/joins/README.md [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+   ```
+
+   This command reates the `readings/ID1234` and  `parameters` directories
+   with the structures described above.
+
+1. Go to [Step 2](#step-2-set-up-the-pachyderm-repositories).
+
+## Step 2 - Set up the Pachyderm Repositories
+
+You need to configure the `readings` and `parameters` repositories
```
Comment (JoeyZwicker) [inline examples/joins/README.md]:
```suggestion
You need to create the `readings` and `parameters` repositories
```

File: examples/joins/README.md [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+    NAME       TYPE SIZE
+    /file1.txt file 9B
+    /file2.txt file 9B
+    /file3.txt file 9B
+    /file4.txt file 9B
+    /file5.txt file 9B
+    /file6.txt file 9B
+    ```
+
+1. From the `readings directory, run the following command to upload
```
Comment (JoeyZwicker) [inline examples/joins/README.md]:
```suggestion
1. From the `readings` directory, run the following command to upload
```

File: examples/joins/parameters/file1.txt [hunk 14]
```diff
@@ -0,0 +1,5 @@
+1
```
Comment (JoeyZwicker) [inline examples/joins/parameters/file1.txt]: you're uploading the dummy files and having them create them with the makefile? Shouldn't we only do one or the other? Or were these pushed by accident?

File: doc/concepts/pipeline-concepts/pipeline/join.md (hunk 10 above)
Comment (svekars) [reply doc/concepts/pipeline-concepts/pipeline/join.md]: adddress

File: doc/concepts/pipeline-concepts/pipeline/join.md (hunk 11 above)
Comment (svekars) [reply doc/concepts/pipeline-concepts/pipeline/join.md]: addressed

File: examples/joins/README.md (hunk 12 above)
Comment (svekars) [reply examples/joins/README.md]: addressed

File: examples/joins/README.md (hunk 13 above)
Comment (svekars) [reply examples/joins/README.md]: addressed

File: examples/joins/parameters/file1.txt (hunk 14 above)
Comment (svekars) [reply examples/joins/parameters/file1.txt]: by accident 🤦‍♀

**Pull request merged**